### PR TITLE
Add support for Date primitive

### PR DESCRIPTION
**Pull request by NancyGargMS**
Added support for Date primitive type.
Extended JsonConverter to support DateOnly conversions (for maps as well).

### CODE REVIEW

File: src/Generator.TemplateProject/Custom/DateOnlyConverter.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="DateOnlyConverter"/> class.
+    /// </summary>
+    /// <param name="serializationFormat">Serialization format. </param>
+    public DateOnlyConverter(string? serializationFormat)
+    {
+        this.serializationFormat = serializationFormat ?? "yyyy-MM-dd";
+    }
```
Comment (jarz) [inline src/Generator.TemplateProject/Custom/DateOnlyConverter.cs:35]: I don't believe we need this constructor. The RFC section specifies the exact format of a full-date string which is what's supported in DTDLv2. Allowing someone to override that seems like a mistake.

File: src/Generator.TemplateProject/Custom/MapDateOnlyConverter.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    {
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="MapDateOnlyConverter"/> class.
+    /// </summary>
+    /// <param name="serializationFormat">Serialization format. </param>
+    public MapDateOnlyConverter(string? serializationFormat)
+    {
+        this.serializationFormat = serializationFormat ?? "yyyy-MM-dd";
```
Comment (jarz) [inline src/Generator.TemplateProject/Custom/MapDateOnlyConverter.cs:34]: See comment on `DateOnlyConverter`.

File: src/Generator/Base/ClassEntity.cs [hunk 3]
```diff
@@ -24,6 +24,7 @@ protected override void WriteUsingStatements(StreamWriter streamWriter)
         WriteUsingAzure(streamWriter);
         WriteUsingAdt(streamWriter);
         WriteUsingSystem(streamWriter);
+        WriteUsingSystemRuntime(streamWriter);
```
Comment (jarz) [inline src/Generator/Base/ClassEntity.cs:27]: Is this necessary? The [DateOnly documentation](https://learn.microsoft.com/en-us/dotnet/api/system.dateonly?view=net-6.0) has the struct in the `System` namespace, not `System.Runtime`.

File: test/Generator.Tests/MapDateOnlyConverter.UnitTests.cs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                ["Friday"] = new DateOnly(2020, 03, 10),
+                ["Sunday"] = new DateOnly(2020, 03, 12)
+            }
+        };
+        var deserializedAsset = JsonSerializer.Deserialize<Asset>(json, options);
+        Assert.AreEqual(expectedAsset.Id, deserializedAsset?.Id);
+        Assert.AreEqual(expectedAsset.Name, deserializedAsset?.Name);
+        Assert.AreEqual(expectedAsset.SerialNumber, deserializedAsset?.SerialNumber);
+        Assert.AreEqual(expectedAsset.RuntimeDetails["Friday"], deserializedAsset?.RuntimeDetails["Friday"]);
+        Assert.AreEqual(expectedAsset.RuntimeDetails["Sunday"], deserializedAsset?.RuntimeDetails["Sunday"]);
```
Comment (jarz) [inline test/Generator.Tests/MapDateOnlyConverter.UnitTests.cs:32]: The null-forgiving operator probably needs to be used here.

**Review by jarz — COMMENTED** [review]
Just a few small changes, looking good otherwise.

File: src/Generator.TemplateProject/Custom/DateOnlyConverter.cs (hunk 1 above)
Comment (NancyGargMS) [reply src/Generator.TemplateProject/Custom/DateOnlyConverter.cs]: done

File: src/Generator.TemplateProject/Custom/MapDateOnlyConverter.cs (hunk 2 above)
Comment (NancyGargMS) [reply src/Generator.TemplateProject/Custom/MapDateOnlyConverter.cs]: Done

File: src/Generator.TemplateProject/Custom/DateOnlyConverter.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace Generator.CustomModels;
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using System.Xml;
+//using System.Runtime;
```
Comment (jarz) [inline src/Generator.TemplateProject/Custom/DateOnlyConverter.cs:12]: I'm sorry - I missed this in the last PR. Otherwise everything looks great!

**Review by jarz — COMMENTED** [review]
One more thing I missed yesterday. Sorry about that!

File: src/Generator.TemplateProject/Custom/MapDateOnlyConverter.cs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+            var key = reader.GetString();
+
+            reader.Read();
+            if (reader.TokenType == JsonTokenType.Null)
+            {
+                continue;
+            }
+
+#pragma warning disable CS8604 // Possible null reference argument.
+            var value = DateOnly.Parse(reader.GetString());
```
Comment (ms-mikeb) [inline src/Generator.TemplateProject/Custom/MapDateOnlyConverter.cs:51]: What it is complaining that it may be null here?

Comment (NancyGargMS) [reply src/Generator.TemplateProject/Custom/MapDateOnlyConverter.cs:51]: It is showing the warnings about null references for "reader" and "key" in the following code:
            var value = DateOnly.Parse(reader.GetString());
            dictionary.Add(key, value);

Comment (jarz) [reply src/Generator.TemplateProject/Custom/MapDateOnlyConverter.cs:51]: I had to make a similar concession in [MapDurationConverter.cs](https://github.com/microsoft/dtdl-model-generator/blob/96e1cae8ddd7ed3b13725bab4bb8003289972d1e/src/Generator.TemplateProject/Custom/MapDurationConverter.cs#L49-L55). I added an additional comment explaining the above check for `JsonTokenType.Null` handles this situation.

File: test/Generator.Tests/MapDateOnlyConverter.UnitTests.cs [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        var json = $"{{\"$dtId\":\"d8985302-4ee1-4a10-b2f5-e854e1682422\",\"assetTag\":\"12345\",\"name\":\"Test Asset\",\"serialNumber\":\"SN12345\",\"runtimeDetails\":{{\"Friday\":\"2020-03-10\",\"Sunday\":\"2020-03-12\"}},\"$metadata\":{{\"$model\":\"{Asset.ModelId}\"}}}}";
+        var expectedAsset = new Asset
+        {
+            Id = "d8985302-4ee1-4a10-b2f5-e854e1682422",
+            AssetTag = "ABC123",
+            Name = "Test Asset",
+            SerialNumber = "SN12345",
+            RuntimeDetails = new Dictionary<string, DateOnly>()
+            {
+                ["Friday"] = new DateOnly(2020, 03, 10),
```
Comment (ms-mikeb) [inline test/Generator.Tests/MapDateOnlyConverter.UnitTests.cs:23]: This is a rather uncommon way to instantiate a key/value pair and isn't consistent with other ways we do it. 

try:
``` csharp
{
    { "Friday", new DateOnly(2020, 03, 10) },
    { "Sunday", new DateOnly(2020, 03, 12) }
}
```

File: test/Generator.Tests/MapDateOnlyConverter.UnitTests.cs [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+                ["Sunday"] = new DateOnly(2020, 03, 12)
+            }
+        };
+        var deserializedAsset = JsonSerializer.Deserialize<Asset>(json, options);
+        Assert.AreEqual(expectedAsset.Id, deserializedAsset?.Id);
+        Assert.AreEqual(expectedAsset.Name, deserializedAsset?.Name);
+        Assert.AreEqual(expectedAsset.SerialNumber, deserializedAsset?.SerialNumber);
+        Assert.AreEqual(expectedAsset.RuntimeDetails["Friday"], deserializedAsset?.RuntimeDetails!["Friday"]);
```
Comment (ms-mikeb) [inline test/Generator.Tests/MapDateOnlyConverter.UnitTests.cs:31]: Rule of thumb for string literals...
1. If you're using the same string literal in multiple files, make it a project/solution-level constant.
2. If you're using the same string literal multiple times in the same file, make it a file-level constant.

File: src/Generator.TemplateProject/Custom/DateOnlyConverter.cs (hunk 5 above)
Comment (ms-mikeb) [inline src/Generator.TemplateProject/Custom/DateOnlyConverter.cs:12]: Remove

File: src/Generator.TemplateProject/Custom/DateOnlyConverter.cs [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+/// Converts an RFC 3339 Date into a DateOnly.
+/// </summary>
+public class DateOnlyConverter : JsonConverter<DateOnly>
+{
+    /// <inheritdoc/>
+    public override DateOnly Read(ref Utf8JsonReader reader,
+                            Type typeToConvert, JsonSerializerOptions options)
+    {
+        var value = reader.GetString();
+        return DateOnly.Parse(value!);
```
Comment (ms-mikeb) [inline src/Generator.TemplateProject/Custom/DateOnlyConverter.cs:24]: Null-Assertions should be used sparingly.  It's telling you that something could be null, so you should be checking if it's null and adding specific behavior/error checking.

File: src/Generator.TemplateProject/Custom/MapDateOnlyConverter.cs (hunk 6 above)
Comment (ms-mikeb) [reply src/Generator.TemplateProject/Custom/MapDateOnlyConverter.cs:51]: I think it'd be better if we just updated the if check then to include `if (reader.TokenType == JsonTokenType.Null || string.IsNullOrWhiteSpace(key))`

Futhermore we should make `var dictionary = new Dictionary<string, DateOnly>();` the first line of code in the method and then add:
``` csharp
if (reader is null)
{
    return dictionary;
}
```

Then we're properly handling what the compiler is suggesting we handle vs just ignoring it.

**Review by ms-mikeb — CHANGES_REQUESTED** [review]

File: src/Generator.TemplateProject/Custom/DateOnlyConverter.cs (hunk 5 above)
Comment (NancyGargMS) [reply src/Generator.TemplateProject/Custom/DateOnlyConverter.cs]: Done

File: src/Generator.TemplateProject/Custom/MapDateOnlyConverter.cs [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Text;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using System.Xml;
+
+/// <summary>
+/// Converts a map of RFC 3339 into an IDictionary&lt;string, DateOnly&lt;.
```
Comment (ms-mikeb) [inline src/Generator.TemplateProject/Custom/MapDateOnlyConverter.cs:15]: I wouldn't suggest trying to put the type in the comment like this. Use a cref instead. `<see cref="IDictionary<string, DateOnly>"/>`

**Review by ms-mikeb — CHANGES_REQUESTED** [review]

**Review by ms-mikeb — APPROVED** [review]

**Review by jarz — APPROVED** [review]

**Pull request merged**